src/dbnode/persist/fs/retriever.go

```diff
@@ -746,7 +766,9 @@ func (reqs *shardRetrieveRequests) resetQueued() {
 
 // Don't forget to update the resetForReuse method when adding a new field
 type retrieveRequest struct {
-	resultWg sync.WaitGroup
+	finalized   bool
+	waitOnFetch chan struct{}
```

Comment:
What was the reason to replace `sync.WaitGroup` with a channel (see my concern about the busy wait above)?

Reply:
The wait group is still present, it's actually used for a slightly different mechanism for synchronizing request reads (this new channel is for ensuring that we correctly call OnRetrieveBlock)